libgadget/forcetree.c

```diff
@@ -402,52 +387,76 @@ int force_tree_create_nodes(const ForceTree tb, const int npart, DomainDecomp *
             const int topleaf = domain_get_topleaf(P[i].Key, ddecomp);
             this = ddecomp->TopLeaves[topleaf].treenode;
         }
-        int child, subnode;
+        int child;
+        int nocc;
 
         /*Walk the main tree until we get something that isn't an internal node.*/
         do
         {
-            /*We will always start with an internal node: find the desired subnode.*/
-            subnode = get_subnode(&tb.Nodes[this], i);
-
             /*No lock needed: if we have an internal node here it will be stable*/
             #pragma omp atomic read
-            child = tb.Nodes[this].u.suns[subnode];
+            nocc = tb.Nodes[this].u.s.noccupied;
+
+            /* This node still has space for a particle (or needs conversion)*/
+            if(nocc < (1 << 16))
+                break;
+
+            /* This node has child subnodes: find them.*/
+            int subnode = get_subnode(&tb.Nodes[this], i);
+            /*No lock needed: if we have an internal node here it will be stable*/
+            child = tb.Nodes[this].u.s.suns[subnode];
 
-            if(child > tb.lastnode)
+            if(child > tb.lastnode || child < tb.firstnode)
                 endrun(1,"Corruption in tree build: N[%d].[%d] = %d > lastnode (%d)\n",this, subnode, child, tb.lastnode);
-            /* If we found an internal node keep walking*/
-            else if(child >= tb.firstnode) {
-                this = child;
-            }
+            this = child;
         }
         while(child >= tb.firstnode);
 
         /*Now lock this node.*/
         lock_spinlock(this-tb.firstnode, spin);
+        /* We have a guaranteed spot.*/
+        nocc = atomic_fetch_and_add(&tb.Nodes[this].u.s.noccupied, 1);
+
+        /* Check whether there is now a new layer of nodes and if so walk down until there isn't.*/
+        if(nocc >= (1<<16)) {
```

Comment:
Is it possible to store the flag of 'have_node_children' and 'have_particle_children' to a more explicit variable? I think adding new member to the node struct is OK now as you are significiantlly reducing the number of nodes by allowing more than 1 particle in a node.

Reply:
I will do this, but I will leave the nocc logic alone as well: it helps simplify the threading.